functions/src/resolve-market.ts

```diff
@@ -31,140 +41,215 @@ export const resolveMarket = functions
     ) => {
       const userId = context?.auth?.uid
       if (!userId) return { status: 'error', message: 'Not authorized' }
-
-      const { outcome, contractId, probabilityInt, resolutions, value } = data
-
-      const contractDoc = firestore.doc(`contracts/${contractId}`)
+      const contractDoc = firestore.doc(`contracts/${data.contractId}`)
       const contractSnap = await contractDoc.get()
       if (!contractSnap.exists)
         return { status: 'error', message: 'Invalid contract' }
       const contract = contractSnap.data() as Contract
-      const { creatorId, outcomeType, closeTime } = contract
-
-      if (outcomeType === 'BINARY') {
-        if (!RESOLUTIONS.includes(outcome))
-          return { status: 'error', message: 'Invalid outcome' }
-      } else if (outcomeType === 'FREE_RESPONSE') {
-        if (
-          isNaN(+outcome) &&
-          !(outcome === 'MKT' && resolutions) &&
-          outcome !== 'CANCEL'
-        )
-          return { status: 'error', message: 'Invalid outcome' }
-      } else if (outcomeType === 'NUMERIC') {
-        if (isNaN(+outcome) && outcome !== 'CANCEL')
-          return { status: 'error', message: 'Invalid outcome' }
-      } else {
-        return { status: 'error', message: 'Invalid contract outcomeType' }
-      }
-
-      if (value !== undefined && !isFinite(value))
-        return { status: 'error', message: 'Invalid value' }
+      if (contract.creatorId !== userId)
+        return { status: 'error', message: 'User not creator of contract' }
+
+      return privateResolveMarket(contract, data)
+    }
+  )
+
+export const autoResolveMarkets = functions
+  .runWith({ secrets: ['MAILGUN_KEY'] })
+  .pubsub.schedule('every 1 minutes')
+  .onRun(async () => {
+    const contracts = await getValues<Contract>(
+      firestore
+        .collection('contracts')
+        .where('isResolved', '==', false)
+        .where('autoResolutionTime', '<', Date.now())
+    )
+
+    await batchedWaitAll(
+      contracts.map((contract) => async () => {
+        await autoResolve(contract)
+      })
+    )
+  })
+
+const autoResolve = async (contract: Contract) => {
+  const { closeTime, autoResolutionTime } = contract
+  if (closeTime && autoResolutionTime && closeTime > autoResolutionTime) {
+    console.error(
+      'contract',
+      contract.id,
+      'is supposed to be resolved before it closes. Close time:',
+      new Date(closeTime),
+      'Auto resolution time:',
+      new Date(autoResolutionTime)
+    )
+    return { status: 'error', message: 'Invalid auto resolution time' }
+  }
```

Comment:
RE: https://github.com/manifoldmarkets/manifold/pull/404#discussion_r897551724
@jahooma this is my suggestion instead of filtering for the close time: close time > auto resolution time is an error and will be logged this way.